network/src/client.rs

```diff
@@ -118,24 +120,34 @@ macro_rules! define_method {
 }
 
 impl Client {
-    pub fn register_extension<T>(&self, extension: Arc<T>)
+    pub fn new_extension<T, F>(&self, factory: F) -> Arc<T>
     where
-        T: 'static + Sized + NetworkExtension, {
-        let name = extension.name();
+        T: 'static + Sized + NetworkExtension,
+        F: FnOnce(&Arc<Api>) -> T, {
```

Comment:
How about moving `Arc<Api>` to factory?

Reply:
`fn new(...) -> Arc<Self>` requires disabling clippy warning, `.new_extension(|api| Arc::new(Extension::new())` is verbose, and there's no use case other than that. And I think it's natural for a container-like data structure to rule over how data is managed.

Reply:
?

Reply:
I misunderstood your comment. Sorry. It was changed.